File: golang/commands.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

// --- List Command ---

// listCommand lists all servers and their tools.
func listCommand(configPath string, withDescriptions bool) {
	config, _, err := loadConfig(configPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	serverNames := listServerNames(config)
	if len(serverNames) == 0 {
		fmt.Fprintln(os.Stderr, "Warning: No servers configured. Add servers to mcp_servers.json")
		return
	}

	sort.Strings(serverNames)
	concurrency := getConcurrencyLimit()

	debugLog("Processing %d servers with concurrency %d", len(serverNames), concurrency)

	servers := fetchAllServerTools(serverNames, config, concurrency)

	// Sort by name
	sort.Slice(servers, func(i, j int) bool {
	return servers[i].name < servers[j].name
	})

	// Convert errors to display format
	var displayServers []ServerWithTools
	for _, s := range servers {
		swt := ServerWithTools{
			Name:         s.name,
			Tools:        s.tools,
			Instructions: s.instructions,
		}
		if s.err != "" {
			swt.Tools = []ToolInfo{{Name: fmt.Sprintf("<error: %s>", s.err)}}
		}
		displayServers = append(displayServers, swt)
	}

	fmt.Println(formatServerList(displayServers, withDescriptions))
}

type serverResult struct {
	name         string
	tools        []ToolInfo
	instructions string
	err          string
}

// fetchAllServerTools fetches tools from all servers with concurrency limit.
func fetchAllServerTools(serverNames []string, config *McpServersConfig, concurrency int) []serverResult {
	results := make([]serverResult, len(serverNames))
	sem := make(chan struct{}, concurrency)
	var wg sync.WaitGroup

	for i, name := range serverNames {
		wg.Add(1)
		go func(idx int, serverName string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			results[idx] = fetchServerTools(serverName, config)
		}(i, name)
	}

	wg.Wait()
	return results
}

// fetchServerTools fetches tools from a single server.
func fetchServerTools(serverName string, config *McpServersConfig) serverResult {
	serverConfig, err := getServerConfig(config, serverName)
	if err != nil {
		return serverResult{name: serverName, err: err.Error()}
	}

	conn, err := getUnifiedConnection(serverName, serverConfig)
	if err != nil {
		debugLog("%s: connection failed - %s", serverName, err.Error())
		return serverResult{name: serverName, err: err.Error()}
	}
	defer conn.Close()

	tools, err := conn.ListTools()
	if err != nil {
		return serverResult{name: serverName, err: err.Error()}
	}

	// Apply tool filtering
	tools = filterTools(tools, serverConfig)

	instructions := conn.GetInstructions()
	debugLog("%s: loaded %d tools", serverName, len(tools))

	return serverResult{
		name:         serverName,
		tools:        tools,
		instructions: instructions,
	}
}

// --- Info Command ---

// infoCommand shows server or tool details.
func infoCommand(serverName, toolName, configPath string, withDescriptions bool) {
	config, _, err := loadConfig(configPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	serverConfig, err := getServerConfig(config, serverName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	conn, err := getUnifiedConnection(serverName, serverConfig)
	if err != nil {
		fmt.Fprintln(os.Stderr, formatCliError(serverConnectionError(serverName, err.Error())))
		os.Exit(ErrorCodeNetworkError)
	}
	defer conn.Close()

	tools, err := conn.ListTools()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to list tools: %s\n", err.Error())
		os.Exit(ErrorCodeServerError)
	}

	// Apply tool filtering
	tools = filterTools(tools, serverConfig)

	if toolName != "" {
		// Show specific tool schema
		var tool *ToolInfo
		for i := range tools {
			if tools[i].Name == toolName {
				tool = &tools[i]
				break
			}
		}

		if tool == nil {
			var available []string
			for _, t := range tools {
				available = append(available, t.Name)
			}
			fmt.Fprintln(os.Stderr, formatCliError(toolNotFoundError(toolName, serverName, available)))
			os.Exit(ErrorCodeClientError)
		}

		fmt.Println(formatToolSchema(serverName, *tool))
	} else {
		// Show server details
		instructions := conn.GetInstructions()
		fmt.Println(formatServerDetails(serverName, serverConfig, tools, withDescriptions, instructions))
	}
}

// --- Grep Command ---

// globToRegex converts a glob pattern to a regex.
func globToRegex(pattern string) *regexp.Regexp {
	var escaped strings.Builder
	i := 0

	for i < len(pattern) {
		ch := pattern[i]

		if ch == '*' && i+1 < len(pattern) && pattern[i+1] == '*' {
			// ** (globstar) - match anything including slashes
			escaped.WriteString(".*")
			i += 2
			for i < len(pattern) && pattern[i] == '*' {
				i++
			}
		} else if ch == '*' {
			// * - match any chars except slash
			escaped.WriteString("[^/]*")
			i++
		} else if ch == '?' {
			// ? - match single char (not slash)
			escaped.WriteString("[^/]")
			i++
		} else if strings.ContainsRune("[.+^${}()|\\]", rune(ch)) {
			escaped.WriteByte('\\')
			escaped.WriteByte(ch)
			i++
		} else {
			escaped.WriteByte(ch)
			i++
		}
	}

	re, err := regexp.Compile("(?i)^" + escaped.String() + "$")
	if err != nil {
		// Fallback to literal match
		re = regexp.MustCompile("(?i)^" + regexp.QuoteMeta(pattern) + "$")
	}
	return re
}

// grepCommand searches tools by pattern across all servers.
func grepCommand(pattern, configPath string, withDescriptions bool) {
	config, _, err := loadConfig(configPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	re := globToRegex(pattern)
	serverNames := listServerNames(config)

	if len(serverNames) == 0 {
		fmt.Fprintln(os.Stderr, "Warning: No servers configured. Add servers to mcp_servers.json")
		return
	}

	concurrency := getConcurrencyLimit()
	debugLog("Searching %d servers for pattern %q (concurrency: %d)", len(serverNames), pattern, concurrency)

	// Search all servers in parallel
	type searchResult struct {
		serverName string
		results    []SearchResult
		err        string
	}

	allSearchResults := make([]searchResult, len(serverNames))
	sem := make(chan struct{}, concurrency)
	var wg sync.WaitGroup

	for i, name := range serverNames {
		wg.Add(1)
		go func(idx int, serverName string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			serverConfig, err := getServerConfig(config, serverName)
			if err != nil {
				allSearchResults[idx] = searchResult{serverName: serverName, err: err.Error()}
				return
			}

			conn, err := getUnifiedConnection(serverName, serverConfig)
			if err != nil {
				debugLog("%s: connection failed - %s", serverName, err.Error())
				allSearchResults[idx] = searchResult{serverName: serverName, err: err.Error()}
				return
			}
			defer conn.Close()

			tools, err := conn.ListTools()
			if err != nil {
				allSearchResults[idx] = searchResult{serverName: serverName, err: err.Error()}
				return
			}

			// Apply tool filtering
			tools = filterTools(tools, serverConfig)

			var results []SearchResult
			for _, tool := range tools {
				if re.MatchString(tool.Name) {
					results = append(results, SearchResult{Server: serverName, Tool: tool})
				}
			}

			debugLog("%s: found %d matches", serverName, len(results))
			allSearchResults[idx] = searchResult{serverName: serverName, results: results}
		}(i, name)
	}

	wg.Wait()

	// Collect results
	var allResults []SearchResult
	var failedServers []string
	for _, sr := range allSearchResults {
		allResults = append(allResults, sr.results...)
		if sr.err != "" {
			failedServers = append(failedServers, sr.serverName)
		}
	}

	if len(failedServers) > 0 {
		fmt.Fprintf(os.Stderr, "Warning: %d server(s) failed to connect: %s\n",
			len(failedServers), strings.Join(failedServers, ", "))
	}

	if len(allResults) == 0 {
		fmt.Printf("No tools found matching %q\n", pattern)
		fmt.Println("  Tip: Pattern matches tool names only (not server names)")
		fmt.Println("  Tip: Use '*' for wildcards, e.g. '*file*' or 'read_*'")
		fmt.Println("  Tip: Run 'mcp-cli' to list all available tools")
		return
	}

	fmt.Println(formatSearchResults(allResults, withDescriptions))
}

// --- Call Command ---

// callCommand executes a tool call.
func callCommand(serverName, toolName, argsStr, configPath string) {
	config, _, err := loadConfig(configPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	serverConfig, err := getServerConfig(config, serverName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	// Parse arguments
	args, err := parseCallArgs(argsStr)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(ErrorCodeClientError)
	}

	// Check if tool is allowed before calling
	if !isToolAllowed(toolName, serverConfig) {
		fmt.Fprintln(os.Stderr, formatCliError(toolDisabledError(toolName, serverName)))
		os.Exit(ErrorCodeClientError)
	}

	conn, err := getUnifiedConnection(serverName, serverConfig)
	if err != nil {
		fmt.Fprintln(os.Stderr, formatCliError(serverConnectionError(serverName, err.Error())))
		os.Exit(ErrorCodeNetworkError)
	}
	defer conn.Close()

	result, err := conn.CallTool(toolName, args)
	if err != nil {
		errMsg := err.Error()

		// Try to get available tools for better error message
		if strings.Contains(errMsg, "not found") || strings.Contains(errMsg, "unknown tool") {
			var available []string
			if tools, listErr := conn.ListTools(); listErr == nil {
				for _, t := range tools {
					available = append(available, t.Name)
				}
			}
			fmt.Fprintln(os.Stderr, formatCliError(toolNotFoundError(toolName, serverName, available)))
		} else {
			fmt.Fprintln(os.Stderr, formatCliError(toolExecutionError(toolName, serverName, errMsg)))
		}
		os.Exit(ErrorCodeServerError)
	}

	fmt.Println(formatToolResult(result))
}

// parseCallArgs parses JSON arguments from string or stdin.
func parseCallArgs(argsStr string) (map[string]interface{}, error) {
	var jsonString string

	if argsStr != "" {
		jsonString = argsStr
	} else {
		// Check if stdin has data (not a terminal)
		stat, _ := os.Stdin.Stat()
		if (stat.Mode() & os.ModeCharDevice) == 0 {
			timeoutMs := getTimeoutMs()
			type readResult struct {
				data []byte
				err  error
			}
			ch := make(chan readResult, 1)
			go func() {
				data, err := io.ReadAll(os.Stdin)
				ch <- readResult{data, err}
			}()
			select {
			case result := <-ch:
				if result.err != nil {
					return nil, fmt.Errorf("failed to read from stdin: %w", result.err)
				}
				jsonString = strings.TrimSpace(string(result.data))
			case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
				return nil, fmt.Errorf("stdin read timed out after %dms", timeoutMs)
			}
		}
	}

	if jsonString == "" {
		return map[string]interface{}{}, nil
	}

	var args map[string]interface{}
	if err := json.Unmarshal([]byte(jsonString), &args); err != nil {
		return nil, invalidJSONArgsError(jsonString, err.Error())
	}

	return args, nil
}

File: rust/src/commands.rs
use std::io::{self, Read};
use std::sync::Arc;

use tokio::sync::Semaphore;

use crate::client::*;
use crate::config::*;
use crate::errors::*;
use crate::output::*;

// --- List Command ---

pub async fn list_command(config_path: Option<&str>, with_descriptions: bool) -> Result<(), CliError> {
    let (config, _) = load_config(config_path)?;
    let mut server_names = list_server_names(&config);
    server_names.sort();

    if server_names.is_empty() {
        eprintln!("Warning: No servers configured. Add servers to mcp_servers.json");
        return Ok(());
    }

    let concurrency = get_concurrency_limit();
    debug_log(&format!("Processing {} servers with concurrency {}", server_names.len(), concurrency));

    let results = fetch_all_server_tools(&server_names, &config, concurrency).await;

    let mut display_servers: Vec<ServerWithTools> = results
        .into_iter()
        .map(|r| {
            let tools = if let Some(ref err) = r.error {
                vec![ToolInfo {
                    name: format!("<error: {}>", err),
                    description: None,
                    input_schema: serde_json::json!({}),
                }]
            } else {
                r.tools
            };
            ServerWithTools {
                name: r.name,
                tools,
                instructions: r.instructions,
                error: r.error,
            }
        })
        .collect();

    display_servers.sort_by(|a, b| a.name.cmp(&b.name));
    println!("{}", format_server_list(&display_servers, with_descriptions));
    Ok(())
}

struct ServerResult {
    name: String,
    tools: Vec<ToolInfo>,
    instructions: Option<String>,
    error: Option<String>,
}

async fn fetch_all_server_tools(
    server_names: &[String],
    config: &McpServersConfig,
    concurrency: usize,
) -> Vec<ServerResult> {
    let sem = Arc::new(Semaphore::new(concurrency));
    let mut handles = Vec::new();

    for name in server_names {
        let name = name.clone();
        let config = config.clone();
        let sem = sem.clone();

        let handle = tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            fetch_server_tools(&name, &config).await
        });
        handles.push(handle);
    }

    let mut results = Vec::new();
    for handle in handles {
        match handle.await {
            Ok(result) => results.push(result),
            Err(e) => results.push(ServerResult {
                name: "unknown".to_string(),
                tools: Vec::new(),
                instructions: None,
                error: Some(e.to_string()),
            }),
        }
    }
    results
}

async fn fetch_server_tools(server_name: &str, config: &McpServersConfig) -> ServerResult {
    let server_config = match get_server_config(config, server_name) {
        Ok(c) => c,
        Err(e) => return ServerResult {
            name: server_name.to_string(),
            tools: Vec::new(),
            instructions: None,
            error: Some(e.message),
        },
    };

    let connection = match get_connection(server_name, server_config).await {
        Ok(c) => c,
        Err(e) => {
            debug_log(&format!("{}: connection failed - {}", server_name, e.message));
            return ServerResult {
                name: server_name.to_string(),
                tools: Vec::new(),
                instructions: None,
                error: Some(e.message),
            };
        }
    };

    let tools = match connection.list_tools().await {
        Ok(t) => t,
        Err(e) => {
            connection.close().await;
            return ServerResult {
                name: server_name.to_string(),
                tools: Vec::new(),
                instructions: None,
                error: Some(e.message),
            };
        }
    };

    let instructions = connection.get_instructions_async().await;
    debug_log(&format!("{}: loaded {} tools", server_name, tools.len()));

    connection.close().await;

    ServerResult {
        name: server_name.to_string(),
        tools,
        instructions,
        error: None,
    }
}

// --- Info Command ---

pub async fn info_command(
    server_name: &str,
    tool_name: Option<&str>,
    config_path: Option<&str>,
    with_descriptions: bool,
) -> Result<(), CliError> {
    let (config, _) = load_config(config_path)?;
    let server_config = get_server_config(&config, server_name)?;

    let connection = get_connection(server_name, server_config).await
        .map_err(|e| {
            eprintln!("{}", format_cli_error(&server_connection_error(server_name, &e.message)));
            e
        })?;

    let tools = match connection.list_tools().await {
        Ok(t) => t,
        Err(e) => {
            connection.close().await;
            return Err(e);
        }
    };

    if let Some(tool) = tool_name {
        let found = tools.iter().find(|t| t.name == tool);
        match found {
            Some(t) => {
                println!("{}", format_tool_schema(server_name, t));
            }
            None => {
                let available: Vec<String> = tools.iter().map(|t| t.name.clone()).collect();
                connection.close().await;
                return Err(tool_not_found_error(tool, server_name, &available));
            }
        }
    } else {
        let instructions = connection.get_instructions_async().await;
        println!("{}", format_server_details(
            server_name,
            server_config,
            &tools,
            with_descriptions,
            instructions.as_deref(),
        ));
    }

    connection.close().await;
    Ok(())
}

// --- Grep Command ---

pub async fn grep_command(
    pattern: &str,
    config_path: Option<&str>,
    with_descriptions: bool,
) -> Result<(), CliError> {
    let (config, _) = load_config(config_path)?;
    let re = glob_to_regex(pattern);
    let server_names = list_server_names(&config);

    if server_names.is_empty() {
        eprintln!("Warning: No servers configured. Add servers to mcp_servers.json");
        return Ok(());
    }

    let concurrency = get_concurrency_limit();
    debug_log(&format!("Searching {} servers for pattern {:?} (concurrency: {})", server_names.len(), pattern, concurrency));

    let results = fetch_all_server_tools(&server_names, &config, concurrency).await;

    let mut all_results = Vec::new();
    let mut failed_servers = Vec::new();

    for r in &results {
        if let Some(ref err) = r.error {
            failed_servers.push(r.name.clone());
            let _ = err;
        }
        for tool in &r.tools {
            if re.is_match(&tool.name) {
                all_results.push(SearchResult {
                    server: r.name.clone(),
                    tool: tool.clone(),
                });
            }
        }
    }

    if !failed_servers.is_empty() {
        eprintln!(
            "Warning: {} server(s) failed to connect: {}",
            failed_servers.len(),
            failed_servers.join(", ")
        );
    }

    if all_results.is_empty() {
        println!("No tools found matching {:?}", pattern);
        println!("  Tip: Pattern matches tool names only (not server names)");
        println!("  Tip: Use '*' for wildcards, e.g. '*file*' or 'read_*'");
        println!("  Tip: Run 'mcp-cli' to list all available tools");
        return Ok(());
    }

    println!("{}", format_search_results(&all_results, with_descriptions));
    Ok(())
}

fn glob_to_regex(pattern: &str) -> regex::Regex {
    let mut escaped = String::new();
    let chars: Vec<char> = pattern.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        let ch = chars[i];
        if ch == '*' && i + 1 < chars.len() && chars[i + 1] == '*' {
            escaped.push_str(".*");
            i += 2;
            while i < chars.len() && chars[i] == '*' {
                i += 1;
            }
        } else if ch == '*' {
            escaped.push_str("[^/]*");
            i += 1;
        } else if ch == '?' {
            escaped.push_str("[^/]");
            i += 1;
        } else if "[.+^${}()|\\]".contains(ch) {
            escaped.push('\\');
            escaped.push(ch);
            i += 1;
        } else {
            escaped.push(ch);
            i += 1;
        }
    }

    regex::Regex::new(&format!("(?i)^{}$", escaped))
        .unwrap_or_else(|_| {
            let literal = regex::escape(pattern);
            regex::Regex::new(&format!("(?i)^{}$", literal)).unwrap()
        })
}

// --- Call Command ---

pub async fn call_command(
    server_name: &str,
    tool_name: &str,
    args_str: Option<&str>,
    config_path: Option<&str>,
) -> Result<(), CliError> {
    let (config, _) = load_config(config_path)?;
    let server_config = get_server_config(&config, server_name)?;

    let args = parse_call_args(args_str)?;

    let connection = get_connection(server_name, server_config).await?;

    match connection.call_tool(tool_name, args).await {
        Ok(result) => {
            println!("{}", format_tool_result(&result));
            connection.close().await;
            Ok(())
        }
        Err(e) => {
            let err_msg = e.message.clone();

            if err_msg.contains("not found") || err_msg.contains("unknown tool") {
                let mut available = Vec::new();
                if let Ok(tools) = connection.list_tools().await {
                    available = tools.iter().map(|t| t.name.clone()).collect();
                }
                connection.close().await;
                Err(tool_not_found_error(tool_name, server_name, &available))
            } else {
                connection.close().await;
                Err(tool_execution_error(tool_name, server_name, &err_msg))
            }
        }
    }
}

fn parse_call_args(args_str: Option<&str>) -> Result<serde_json::Map<String, serde_json::Value>, CliError> {
    let json_string = if let Some(s) = args_str {
        if s.is_empty() {
            return Ok(serde_json::Map::new());
        }
        s.to_string()
    } else if !atty::is(atty::Stream::Stdin) {
        let mut buf = String::new();
        io::stdin().read_to_string(&mut buf)
            .map_err(|e| CliError {
                code: ERROR_CODE_CLIENT_ERROR,
                error_type: "STDIN_READ_ERROR".to_string(),
                message: format!("Failed to read from stdin: {}", e),
                details: None,
                suggestion: None,
            })?;
        let trimmed = buf.trim().to_string();
        if trimmed.is_empty() {
            return Ok(serde_json::Map::new());
        }
        trimmed
    } else {
        return Ok(serde_json::Map::new());
    };

    let parsed: serde_json::Value = serde_json::from_str(&json_string)
        .map_err(|e| invalid_json_args_error(&json_string, &e.to_string()))?;

    match parsed.as_object() {
        Some(map) => Ok(map.clone()),
        None => Err(invalid_json_args_error(&json_string, "Expected a JSON object")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_glob_to_regex_exact() {
        let re = glob_to_regex("read_file");
        assert!(re.is_match("read_file"));
        assert!(re.is_match("READ_FILE"));
        assert!(!re.is_match("write_file"));
    }

    #[test]
    fn test_glob_to_regex_star() {
        let re = glob_to_regex("*file*");
        assert!(re.is_match("read_file"));
        assert!(re.is_match("file_read"));
        assert!(re.is_match("myfile"));
        assert!(!re.is_match("directory"));
    }

    #[test]
    fn test_glob_to_regex_prefix() {
        let re = glob_to_regex("read_*");
        assert!(re.is_match("read_file"));
        assert!(re.is_match("read_dir"));
        assert!(!re.is_match("write_file"));
    }

    #[test]
    fn test_glob_to_regex_question_mark() {
        let re = glob_to_regex("file?");
        assert!(re.is_match("file1"));
        assert!(re.is_match("fileA"));
        assert!(!re.is_match("file10"));
        assert!(!re.is_match("file"));
    }

    #[test]
    fn test_glob_to_regex_double_star() {
        let re = glob_to_regex("**");
        assert!(re.is_match("anything"));
        assert!(re.is_match("path/to/file"));
    }

    #[test]
    fn test_glob_to_regex_star_no_slash() {
        let re = glob_to_regex("*");
        assert!(re.is_match("file"));
        assert!(!re.is_match("path/file"));
    }

    #[test]
    fn test_glob_to_regex_special_chars() {
        let re = glob_to_regex("test.file");
        assert!(re.is_match("test.file"));
        assert!(!re.is_match("testXfile"));
    }

    #[test]
    fn test_parse_call_args_empty() {
        let result = parse_call_args(None).unwrap();
        assert!(result.is_empty());
    }

    #[test]
    fn test_parse_call_args_valid_json() {
        let result = parse_call_args(Some(r#"{"path": "./file.txt"}"#)).unwrap();
        assert_eq!(result.get("path").unwrap().as_str(), Some("./file.txt"));
    }

    #[test]
    fn test_parse_call_args_invalid_json() {
        let err = parse_call_args(Some("not json")).unwrap_err();
        assert_eq!(err.error_type, "INVALID_JSON_ARGUMENTS");
    }

    #[test]
    fn test_parse_call_args_empty_string() {
        let result = parse_call_args(Some("")).unwrap();
        assert!(result.is_empty());
    }
}
